File: src/VBNetFx/FacFunctions/Dialogs/EmployeeInfoRelated/frmEmpoyeeHandicapAddEditView.vb
Imports ActiveDev
Imports System.Windows.Forms

Public Class frmEmpoyeeHandicapAddEditView

    ' TODO: Code zum Durchführen der benutzerdefinierten Authentifizierung mithilfe des angegebenen Benutzernamens und des Kennworts hinzufügen 
    ' (Siehe http://go.microsoft.com/fwlink/?LinkId=35339).  
    ' Der benutzerdefinierte Prinzipal kann anschließend wie folgt an den Prinzipal des aktuellen Threads angefügt werden: 
    '     My.User.CurrentPrincipal = CustomPrincipal
    ' wobei CustomPrincipal die IPrincipal-Implementierung ist, die für die Durchführung der Authentifizierung verwendet wird. 
    ' Anschließend gibt My.User Identitätsinformationen zurück, die in das CustomPrincipal-Objekt gekapselt sind,
    ' z.B. den Benutzernamen, den Anzeigenamen usw.

    Private Sub OK_Click(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Me.Close()
    End Sub

    Private Sub Cancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Me.Close()
    End Sub

    Private Sub btnCancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancel.Click
        Me.DialogResult = Windows.Forms.DialogResult.Cancel
    End Sub

    Private Sub btnOk_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnOk.Click

        Try
            ValidateForm()
        Catch ex As InvalidDataException
            MessageBox.Show(ex.Message, "Hinweis")
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Fehler")
            Return
        End Try

        myRow2Edit("ValidFrom") = dtpValidFrom.Value
        ' Double - Parsen : Das geht ohne TryParse, da der Wert bereits in ValidateFrom geprüft wurde
        myRow2Edit("Handicap") = Double.Parse(tbHandicap.Text)

        Me.DialogResult = Windows.Forms.DialogResult.OK
    End Sub


    Private Function GetDateFromRow(ByVal row As DataRow, ByVal column As String) As Date?
        If row(column).GetType Is GetType(DBNull) Then
            Return Nothing
        ElseIf row(column).GetType Is GetType(Date) Then
            Return CDate(row(column))
        End If
    End Function


    Private myRow2Edit As DataRow
    Private myAllRows As DataRowCollection
    Overloads Function ShowDialog(ByVal empDisplayName As String, ByVal row2Edit As DataRow, ByVal allRows As DataRowCollection) As System.Windows.Forms.DialogResult
        lblEmployee.Text = empDisplayName
        myRow2Edit = row2Edit
        myAllRows = allRows
        If GetDateFromRow(myRow2Edit, "ValidFrom").HasValue Then
            dtpValidFrom.Value = GetDateFromRow(myRow2Edit, "ValidFrom").Value
        Else
            ' neuer Datensatz
            dtpValidFrom.Value = Date.Now
        End If

        If row2Edit("Handicap").GetType Is GetType(DBNull) Then
            ' neuer Datensatz
            tbHandicap.Text = "0"
        Else
            tbHandicap.Text = row2Edit("Handicap").ToString
        End If

        Return MyBase.ShowDialog()
    End Function

    Private Sub ValidateForm()
        Dim newHandicap As Double = 0
        Dim handicapError As Boolean = Not Double.TryParse(tbHandicap.Text, newHandicap)
        If Not handicapError Then
            If newHandicap < 0 Then
                handicapError = True
            End If
        End If
        If handicapError Then
            Throw New InvalidDataException("Der angegebene Wert für das Handicap ist ungültig. Er muss numerisch und positiv sein.")
        End If

        Dim newValidFrom As Date = dtpValidFrom.Value.Date
        For Each row As DataRow In myAllRows
            If row IsNot myRow2Edit Then
                If newValidFrom = CDate(row("validfrom")) Then
                    Throw New InvalidDataException("Es existiert bereits ein Handicap Eintrag für den " & newValidFrom.ToString)
                End If
            End If
        Next

    End Sub

    Private Class InvalidDataException
        Inherits Exception
        Public Sub New(ByVal msg As String)
            MyBase.New(msg)
        End Sub
    End Class
End Class

File: src/NetFxCSharp/Facesso.Functions/Dialogs/EmployeeInfoRelated/frmEmpoyeeHandicapAddEditView.cs
using ActiveDev;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Facesso.Functions
{
    public partial class frmEmpoyeeHandicapAddEditView
    {
        // TODO: Code zum Durchführen der benutzerdefinierten Authentifizierung mithilfe des angegebenen Benutzernamens und des Kennworts hinzufügen
        // (Siehe http://go.microsoft.com/fwlink/?LinkId=35339).
        // Der benutzerdefinierte Prinzipal kann anschließend wie folgt an den Prinzipal des aktuellen Threads angefügt werden:
        //     My.User.CurrentPrincipal = CustomPrincipal
        // wobei CustomPrincipal die IPrincipal-Implementierung ist, die für die Durchführung der Authentifizierung verwendet wird.
        // Anschließend gibt My.User Identitätsinformationen zurück, die in das CustomPrincipal-Objekt gekapselt sind,
        // z.B. den Benutzernamen, den Anzeigenamen usw.
        private void OK_Click(System.Object sender, System.EventArgs e)
        {
            this.Close();
        }

        private void Cancel_Click(System.Object sender, System.EventArgs e)
        {
            this.Close();
        }

        private void btnCancel_Click(System.Object sender, System.EventArgs e)
        {
            this.DialogResult = System.Windows.Forms.DialogResult.Cancel;
        }

        private void btnOk_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                ValidateForm();
            }
            catch (InvalidDataException ex)
            {
                MessageBox.Show(ex.Message, "Hinweis");
                return;
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Fehler");
                return;
            }

            myRow2Edit["ValidFrom"] = dtpValidFrom.Value;
            // Double - Parsen : Das geht ohne TryParse, da der Wert bereits in ValidateFrom geprüft wurde
            myRow2Edit["Handicap"] = double.Parse(tbHandicap.Text);
            this.DialogResult = System.Windows.Forms.DialogResult.OK;
        }

        private System.DateTime? GetDateFromRow(DataRow row, string column)
        {
            if (row[column].GetType() == typeof(DBNull))
            {
                return null;
            }
            else if (row[column].GetType() == typeof(System.DateTime))
            {
                return System.Convert.ToDateTime(row[column]);
            }

            return default(System.DateTime? );
        }

        private DataRow myRow2Edit;
        private DataRowCollection myAllRows;
        public System.Windows.Forms.DialogResult ShowDialog(string empDisplayName, DataRow row2Edit, DataRowCollection allRows)
        {
            lblEmployee.Text = empDisplayName;
            myRow2Edit = row2Edit;
            myAllRows = allRows;
            if (GetDateFromRow(myRow2Edit, "ValidFrom").HasValue)
            {
                dtpValidFrom.Value = GetDateFromRow(myRow2Edit, "ValidFrom").Value;
            }
            else
            {
                // neuer Datensatz
                dtpValidFrom.Value = System.DateTime.Now;
            }

            if (row2Edit["Handicap"].GetType() == typeof(DBNull))
            {
                // neuer Datensatz
                tbHandicap.Text = "0";
            }
            else
            {
                tbHandicap.Text = row2Edit["Handicap"].ToString();
            }

            return base.ShowDialog();
        }

        private void ValidateForm()
        {
            double newHandicap = 0;
            bool handicapError = !(double.TryParse(tbHandicap.Text, out newHandicap));
            if (!(handicapError))
            {
                if (newHandicap < 0)
                {
                    handicapError = true;
                }
            }

            if (handicapError)
            {
                throw new InvalidDataException("Der angegebene Wert für das Handicap ist ungültig. Er muss numerisch und positiv sein.");
            }

            System.DateTime newValidFrom = dtpValidFrom.Value.Date;
            foreach (DataRow row in myAllRows)
            {
                if (row != myRow2Edit)
                {
                    if (newValidFrom == System.Convert.ToDateTime(row["validfrom"]))
                    {
                        throw new InvalidDataException("Es existiert bereits ein Handicap Eintrag für den " + newValidFrom.ToString());
                    }
                }
            }
        }

        private class InvalidDataException : Exception
        {
            public InvalidDataException(string msg) : base(msg)
            {
            }
        }

        public frmEmpoyeeHandicapAddEditView()
        {
            InitializeComponent();
        }
    }
}